### PR TITLE
MLC - Levitation Control Logic

### PR DESCRIPTION
**Pull request by jpfbastos**
Lev Control implementation

Missing:
- [x]  PWM Output
- [ ] How to take input
- [ ] PID Constants
- [x] Timer for `dt` in `pid.update`

### CODE REVIEW

**Review by davidbeechey — APPROVED** [review]

File: boards/stm32l432kc/src/bin/levitation.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+use embassy_time::Instant;
+
+use {defmt_rtt as _, panic_probe as _};
+
+const MAX_VOLTAGE: f32 = 500.0; // TBD
+const MAX_CURRENT: f32 = 500.0; // TBD
+const TARGET_HEIGHT: f32 = 10.0; // to be determined by levitation
+
+const GAIN_HEIGHT: PidGain = PidGain{ 
+    // to be determined by levitation
```
Comment (TomLonergan03) [inline boards/stm32l432kc/src/bin/levitation.rs]: `TODOLater` this

File: boards/stm32l432kc/src/bin/levitation.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    kp: 1.0,
+    ki: 0.05,
+    kd: 0.005,
+};
+
+const GAIN_CURRENT: PidGain = PidGain{ 
+    // to be determined by levitation
```
Comment (TomLonergan03) [inline boards/stm32l432kc/src/bin/levitation.rs]: `TODOLater`

File: boards/stm32l432kc/src/bin/levitation.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    let mut pwm = SimplePwm::new(p.TIM2, None, Some(green_light), None, None, hz(2000), CountingMode::EdgeAlignedUp);
+    pwm.enable(Channel::Ch2);
+
+    let max_duty = pwm.get_max_duty() as f32;
+
+    let mut time_start = Instant::now().as_micros() as f32;
+
+    loop {
+        
+        let actual_height = 0.7; // we'll get that from a sensor
```
Comment (TomLonergan03) [inline boards/stm32l432kc/src/bin/levitation.rs]: `TODOLater`

File: boards/stm32l432kc/src/bin/levitation.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    let mut pid_current = Pid::new(GAIN_CURRENT.into());
+
+    let p = embassy_stm32::init(Default::default());
+
+    let green_light = PwmPin::new_ch2(p.PB3, OutputType::PushPull);
+
+    let mut pwm = SimplePwm::new(p.TIM2, None, Some(green_light), None, None, hz(2000), CountingMode::EdgeAlignedUp);
+    pwm.enable(Channel::Ch2);
+
+    let max_duty = pwm.get_max_duty() as f32;
```
Comment (TomLonergan03) [inline boards/stm32l432kc/src/bin/levitation.rs:48]: Factor out the pwm stuff into a `HypedPwm` in `io/pwm.rs` and use that here, so everyone can use PWM as needed

File: boards/stm32l432kc/src/bin/levitation.rs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        
+        let actual_height = 0.7; // we'll get that from a sensor
+        
+        let dt = (Instant::now().as_micros() as f32) - time_start;
+
+        let output_height_pid = pid_height.update(TARGET_HEIGHT, actual_height, dt);
+
+        let mut target_current = actual_height + output_height_pid;
+
+        if target_current >= MAX_CURRENT {target_current = MAX_CURRENT};
```
Comment (TomLonergan03) [inline boards/stm32l432kc/src/bin/levitation.rs]:
```rs
        let target_current = (actual_height + output_height_pid).min(MAX_CURRENT);
```

File: boards/stm32l432kc/src/bin/levitation.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+        let actual_current = 1.0; // we'll get that from a sensor
+
+        let output_current_pid = pid_current.update(target_current, actual_current, dt);
+
+        let mut required_voltage = actual_current + output_current_pid;
+
+        if required_voltage >= MAX_VOLTAGE {required_voltage = MAX_VOLTAGE};
```
Comment (TomLonergan03) [inline boards/stm32l432kc/src/bin/levitation.rs]:
```rs
        let mut required_voltage = (actual_height + output_height_pid).min(MAX_VOLTAGE);
```

File: boards/stm32l432kc/src/bin/levitation.rs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        
+        let dt = (Instant::now().as_micros() as f32) - time_start;
+
+        let output_height_pid = pid_height.update(TARGET_HEIGHT, actual_height, dt);
+
+        let mut target_current = actual_height + output_height_pid;
+
+        if target_current >= MAX_CURRENT {target_current = MAX_CURRENT};
+
+        let actual_current = 1.0; // we'll get that from a sensor
```
Comment (TomLonergan03) [inline boards/stm32l432kc/src/bin/levitation.rs]: `TODOLater`

**Review by TomLonergan03 — APPROVED** [review]

**Review by TomLonergan03 — APPROVED** [review]

**Pull request merged**